src/viadot/sources/bigquery.py

```diff
@@ -0,0 +1,242 @@
+            date_column_name (str, optional): The user can provide the name of the date.
+                If the user-specified column does not exist, all data will be retrieved
+                from the table. Defaults to None.
+            start_date (str, optional): Parameter to pass start date e.g.
+                "2022-01-01". Defaults to None.
+            end_date (str, optional): Parameter to pass end date e.g.
+                "2022-01-01". Defaults to None.
+            columns (list[str], optional): List of columns from given table name.
+                Defaults to None.
+        """
+        if query == "tables":
+            query = self._list_tables(dataset_name=dataset_name)
+
+        elif query == "datasets":
+            query = self._list_datasets()
+
+        else:
+            if date_column_name or columns:
+                table_columns = self._list_columns(
+                    dataset_name=dataset_name, table_name=table_name
+                )
+
+            if columns is None or not set(columns).issubset(set(table_columns)):
+                self.logger.warning(
+                    "Some of the columns provided are either, not in the table or "
+                    + "the list is empty. Downloading all the data instead."
+                )
+                columns = "*"
+            else:
+                columns = ", ".join(columns)
+
+            if date_column_name:
+                if (
+                    not set([date_column_name]).issubset(set(table_columns))
+                    or start_date is None
+                    or end_date is None
+                ):
+                    self.logger.warning(
+                        f"'{date_column_name}' column is not recognized, "
+                        + f"or either `start_date`: {start_date} or `end_date`: "
+                        + f"{end_date} is None.\n"
+                        + "Downloading all the data instead."
+                    )
+                    query = None
+                else:
+                    self.logger.info(
+                        f"Filtering data from date {start_date} to {end_date}"
+                    )
+                    query = f"""
+                        SELECT {columns}
+                            FROM `{self.project_id}.{dataset_name}.{table_name}`
+                            WHERE {date_column_name} BETWEEN
+                                PARSE_DATE("%Y-%m-%d", "{start_date}") AND
+                                PARSE_DATE("%Y-%m-%d", "{end_date}")
+                            ORDER BY {date_column_name} DESC
+                    """
+
+            if query is None:
+                query = (
+                    f"SELECT {columns} "
+                    + f"FROM `{self.project_id}.{dataset_name}.{table_name}`"
+                )
+
+        self.df_data = self._gbd(query)
+
+        self.logger.info(f"Downloaded the data from the table name: '{table_name}'.")
```

Comment:
It's connecting and getting the response, there is another method to return the dataframe.

Reply:
Yes, if you want to keep it that way, it should be renamed to `to_records()`, since `api_connection()` suggests you're creating and likely returning an API connection, which is not true (it downloads data and doesn't return anything, which is highly confusing).

It might be simpler to just remove this method and keep this logic in `to_df()`, as I suggested above, but either way is fine - for sure keeping this as `api_connection()` and not returning the data it's downloading is not good.

Reply:
backlog 2.0 issue:
#1049